clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -580,20 +595,33 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
         return new ClusterAndWaitTime(cluster, elapsed);
     }
 
+    private long waitOnPid(long maxWaitMs) throws InterruptedException {
+        if (transactionState == null) {
```

Comment:
Either that or use `maybeWaitOnPid` to emphasize the fact that it doesn't necessarily involve waiting. Both options work for me.